### ISSUE TITLE
Got `Cannot change thread mode after it is set` on windows

### ISSUE THREAD

**Issue by slmjkdbtl** [tomaka/rodio#214]
```rust
let device = rodio::default_output_device().unwrap();
let source = rodio::Decoder::new(Cursor::new(&include_bytes!("pop.ogg")[..])).unwrap();
rodio::play_raw(&device, source.convert_samples());
```
running a cross-compiled .exe on a virtual box windows (sound card and everything is working fine), got
```
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: -2147417850, kind: Other, message: "Cannot change thread mode after it is set."  }', libcore/result.rs:1009:5

**Comment by 17cupsofcoffee** [comment]
Are you using SDL, by any chance? I ran into this exact same issue, and it seemed to go away when I moved the call to `rodio::default_output_device()` to before SDL got initialized.

In particular, it seemed to start falling over for me any time I initialized Rodio after SDL's game controller module, which seems... strange, and probably coincidental.

If you're not using SDL, hopefully that at least gives you an idea of how you might be able to work around your issue.

**Comment by Xaeroxe** [comment]
I've had this problem with SDL too

**Comment by slmjkdbtl** [comment]
@17cupsofcoffee Yes, I'm using SDL and solved with moving the call before SDL initialization, thanks for the solution, valid workaround for now.

**Comment by 17cupsofcoffee** [comment]
If anyone else runs into this issue, make sure you move Rodio's initialization all the way before SDL, not just early enough into SDL's initialization that it starts working - the latter seems like a valid workaround until [someone turns on some extra settings and it breaks again](https://github.com/17cupsofcoffee/tetra/issues/86) 😅

**Comment by LaineZ** [comment]
i got the same problem while use rodio functions in rust dll lib and call it from 
```
        [DllImport("djmadest_api.dll")]
        private static extern IntPtr init_default_device();
        [DllImport("djmadest_api.dll")]
        private static extern void play_from_file(IntPtr sink, IntPtr filename);
        // in main class
        IntPtr device = init_default_device();
        IntPtr filename = Marshal.StringToHGlobalAnsi("music.mp3");
```

**Comment by woelper** [comment]
I get a similar error `thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: -2147417850, kind: Other, message: "Cannot change thread mode after it is set." }', C:\Users\woelper\.cargo\registry\src\github.com-1ecc6299db9ec823\cpal-0.10.0\src\host\wasapi\com.rs:13:9` as soon as I use it along with imgui-rs (imgui-glium-renderer + glium + imgui-winit-support + winit as recommended).

**Comment by tiborsaas** [comment]
@woelper did you manage to solve it? I tried it with glium too, but failed to make a sound. If I moved the playback to an async function it didn't work, new thread didn't work, moving it to top of `main` also fail :)

**Comment by woelper** [comment]
> 
> 
> @woelper did you manage to solve it? I tried it with glium too, but failed to make a sound. If I moved the playback to an async function it didn't work, new thread didn't work, moving it to top of `main` also fail :)

Well, I did manage to work around it - I used imgui_gfx_renderer and it worked. If you have a similar issue with imgui-rs and rodio you can try that too. If it helps you can look at the support module I am using here:
https://git.schdbr.de/woelper/sampler/src/master/src/support_ogl/mod.rs

**Comment by tlent** [comment]
I ran into this issue as well and found these related issues:
https://github.com/RustAudio/cpal/pull/330
https://github.com/rust-windowing/winit/issues/1185

It seems this happens with versions of winit starting with 0.20. The easiest workaround is to spawn a separate thread for Rodio.

**Comment by Boscop** [comment]
I'm also running into this issue on Win 8.1, I'm not using SDL but glium.
> thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: -2147417850, kind: Other, message: "Cannot change thread mode after it is set." }', C:\Users\me\.cargo\registry\src\github.com-1ecc6299db9ec823\cpal-0.11.0\src\host\wasapi\com.rs:13:9

**Comment by Boscop** [comment]
Hm, with this workaround I can't pause the playback:
```rust
let source = Decoder::new(BufReader::new(File::open(in_path_audio)?))?;
thread::Builder::new()
    .spawn(move || {
        let device = rodio::default_output_device().unwrap();
        rodio::play_raw(&device, source.convert_samples());
    })
    .unwrap();
```

When I try this, it plays no sound:
```rust
let source = BufReader::new(File::open(in_path_audio)?);
let playback_sink = thread::Builder::new()
    .spawn(move || {
        let device = rodio::default_output_device().unwrap();
        rodio::play_once(&device, source)
    })
    .expect("spawn")
    .join()
    .expect("join")?;
playback_sink.play();
```

With this it also plays no sound:
```rust
thread::Builder::new()
    .spawn(move || {
        let device = rodio::default_output_device().unwrap();
        rodio::play_once(&device, source).unwrap();
    })
    .expect("spawn")
    .join()
    .expect("join");
```

Only with this it plays the sound, but then I can't pause/resume the sink:
```rust
thread::Builder::new()
    .spawn(move || {
        let device = rodio::default_output_device().unwrap();
        rodio::play_once(&device, source).unwrap().detach();
    })
    .expect("spawn")
    .join()
    .expect("join");
```

Why doesn't it play any sound unless I detach the sink?

How can I keep the sink so that I can pause/resume it but so that it actually plays audio?

**Comment by tlent** [comment]
Here's what I did:

```Rust
use std::sync::mpsc::{self, Sender};
use std::thread;

use rodio::Sink;

pub struct Audio {
    control_channel: Sender<bool>,
}

impl Audio {
    pub fn new() -> Self {
        let audio_source = rodio::source::SineWave::new(1024);
        let (tx, rx) = mpsc::channel();
        thread::spawn(move || {
            let device = rodio::default_output_device().unwrap();
            let sink = Sink::new(&device);
            sink.pause();
            sink.append(audio_source);
            while let Ok(should_play) = rx.recv() {
                if should_play {
                    sink.play();
                } else {
                    sink.pause();
                }
            }
        });
        Self {
            control_channel: tx,
        }
    }

    pub fn play(&self) {
        self.control_channel.send(true).unwrap();
    }

    pub fn pause(&self) {
        self.control_channel.send(false).unwrap();
    }
}
```

**Comment by bjadamson** [comment]
I am using Glium and Rodio together in my project, I don't need drag-and-drop so I was able to work around this issue by disabling drag and drop support in glium [per changelog](https://github.com/rust-windowing/winit/blob/master/CHANGELOG.md#0221-2020-04-16).

I'm still learning rust, so I'm not in a position to know whether this issue is a glium issue, or a rodio issue.
